### PR TITLE
lrs: handle multiple clusters in LRS stream

### PR DESCRIPTION
**Pull request by menghanl**
- xdsclient.ReportLoad
  - create one LRS stream and one load.Store for each server address
  - to take just the server address, and return the load.Store
  - update EDS and LRS balancing policy to recreate LRS stream only when server address changes
- LRS stream (v2 and v3)
  - set feature `send_all_clusters` in `node`
  - handle `resp.Clusters` and only send load for those clusters
  - handle `resp.SendAllClusters` and send load for all clusters

### CODE REVIEW

**Review by easwars — CHANGES_REQUESTED** [review]
I haven't added much comments. But based on our discussion offline to have the `lrsClient` contain all LRS functionality, I will wait for that change to be made before adding other comments. Thanks.

File: xds/internal/client/client.go [hunk 1]
```diff
@@ -30,6 +30,7 @@ import (
 	v2corepb "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
 	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
 	"github.com/golang/protobuf/proto"
+	load2 "google.golang.org/grpc/xds/internal/client/load"
```
[the remark this answers is not in the document]
Comment (menghanl) [reply xds/internal/client/client.go]: Done

File: xds/internal/client/client_loadreport.go [hunk 2]
```diff
@@ -22,30 +22,46 @@ import (
 
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/xds/internal/client/load"
+	load2 "google.golang.org/grpc/xds/internal/client/load"
```
[the remark this answers is not in the document]
Comment (menghanl) [reply xds/internal/client/client_loadreport.go]: Done

File: xds/internal/client/transport_helper.go [hunk 3]
```diff
@@ -24,6 +24,7 @@ import (
 	"time"
 
 	"github.com/golang/protobuf/proto"
+	load2 "google.golang.org/grpc/xds/internal/client/load"
```
[the remark this answers is not in the document]
Comment (menghanl) [reply xds/internal/client/transport_helper.go]: Done

**Comment by menghanl** [comment]
`lrsClient` done. PTAL.

File: xds/internal/client/client_loadreport_test.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	if u, err := fs2.LRSRequestChan.Receive(ctx); err != nil {
+		t.Errorf("unexpected timeout: %v, %v, want NewConn", u, err)
+	}
+
+	// Send one resp to the client.
+	fs2.LRSResponseChan <- &fakeserver.Response{
+		Resp: &lrspb.LoadStatsResponse{SendAllClusters: true},
+	}
+	// Wait for the send to finish, so server will wait for the next request.
+	time.Sleep(defaultTestShortTimeout)
```
Comment (easwars) [inline xds/internal/client/client_loadreport_test.go:117]: Is there a way to avoid this sleep and instead gate on some event?

**Review by easwars — CHANGES_REQUESTED** [review]

File: xds/internal/client/client_loadreport_test.go (hunk 4 above)
Comment (menghanl) [reply xds/internal/client/client_loadreport_test.go:117]: Done. Not wait for one report and then cancel.

**Review by menghanl — COMMENTED** [review]
Thanks for the review. All fixed. PTAL.

**Comment by menghanl** [comment]
Github PRs are having problems (https://www.githubstatus.com/), and the commit is not picked up by this PR.

File: xds/internal/balancer/edsbalancer/xds_client_wrapper.go [hunk 5]
```diff
@@ -102,7 +101,12 @@ type xdsClientWrapper struct {
 	// xdsClient could come from attributes, or created with balancerName.
 	xdsClient xdsClientInterface
 
-	load *loadStoreWrapper
+	// loadWrapper is a wrapper with loadOriginal, with clusterName and
+	// edsServiceName. It's used children to report loads.
+	loadWrapper *loadStoreWrapper
+	// loadOriginal is the load.Store for reporting loads to lrsServerName. It's
+	// returned by the client.
+	loadOriginal *load.Store
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/edsbalancer/xds_client_wrapper.go:110]: We could instead cache the `load.Store` returned by the call to `xdsclient.ReportLoad()` in the `loadWrapper` itself. And split the `update()` method into two: `setStore()` and `setServiceName()` which will called appropriately when handling the update. What do you think about this approach?

File: xds/internal/balancer/lrs/balancer.go [hunk 6]
```diff
@@ -244,34 +254,38 @@ func (w *xdsClientWrapper) update(newConfig *lbConfig, attr *attributes.Attribut
...
 			w.cancelLoadReport()
 			w.cancelLoadReport = nil
 		}
 		if w.c != nil {
-			w.cancelLoadReport = w.c.ReportLoad(w.lrsServerName, w.clusterName)
+			w.loadOriginal, w.cancelLoadReport = w.c.ReportLoad(w.lrsServerName)
+		}
+	}
+
+	if updateLoadStore {
```
Comment (easwars) [inline xds/internal/balancer/lrs/balancer.go:268]: Probably a copy-paste gotcha. You have the same conditional statement twice.

File: xds/internal/balancer/lrs/balancer.go [hunk 7]
```diff
@@ -199,31 +200,40 @@ type xdsClientWrapper struct {
...
-			if w.c != clientFromAttr {
-				// xds client is different, restart.
-				restartLoadReport = true
-				updateLoadStore = true
-				w.c = clientFromAttr
-			}
-		}
+
+	if attr == nil {
+		return fmt.Errorf("failed to get xdsClient from attributes: attributes is nil")
```
Comment (easwars) [inline xds/internal/balancer/lrs/balancer.go:226]: Should these errors be prefixed with `xds:` or `lrs:`. I'm a little confused about the policy for these error message. I guess we dont have to do that for log statements since the prefix logger takes care of it.

File: xds/internal/client/client_loadreport.go [hunk 8]
```diff
@@ -19,58 +19,123 @@ package client
...
-		dopts := append([]grpc.DialOption{c.opts.Config.Creds}, c.opts.DialOpts...)
-		ccNew, err := grpc.Dial(server, dopts...)
+		lrsC.parent.logger.Infof("LRS server is different from xDS server, starting a new ClientConn")
+		dopts := append([]grpc.DialOption{lrsC.parent.opts.Config.Creds}, lrsC.parent.opts.DialOpts...)
+		ccNew, err := grpc.Dial(lrsC.server, dopts...)
 		if err != nil {
 			// An error from a non-blocking dial indicates something serious.
-			c.logger.Infof("xds: failed to dial load report server {%s}: %v", server, err)
-			return func() {}
+			lrsC.parent.logger.Infof("xds: failed to dial load report server {%s}: %v", lrsC.server, err)
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/client/client_loadreport.go:126]: Hmm ... but we are only throwing an `Info` log here. Am I missing something?

File: xds/internal/client/client_loadreport_test.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if u, err := fs.NewConnChan.Receive(ctx); err != nil {
+		t.Errorf("unexpected timeout: %v, %v, want NewConn", u, err)
+	}
+
+	// Report to the same address should not create new ClientConn.
+	store1, lrsCancel1 := xdsC.ReportLoad(fs.Address)
+	defer lrsCancel1()
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
```
Comment (easwars) [inline xds/internal/client/client_loadreport_test.go:84]: Could you please create a different context here, so that we can use the context with the `defaultTestTimeout` deadline for other things which need it.

```
	sCtx, sCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
	defer sCancel()
	if u, err := fs.NewConnChan.Receive(sCtx); err != context.DeadlineExceeded {
```

File: xds/internal/client/client_loadreport_test.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+	fs2, sCleanup2, err := fakeserver.StartServer()
+	if err != nil {
+		t.Fatalf("failed to start fake xDS server: %v", err)
+	}
+	defer sCleanup2()
+
+	// Report to a different address should create new ClientConn.
+	store2, lrsCancel2 := xdsC.ReportLoad(fs2.Address)
+	defer lrsCancel2()
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
```
Comment (easwars) [inline xds/internal/client/client_loadreport_test.go:99]: If we do the above, we wont need to reinitialize the context with the `defaultTestTimeout` here.

File: xds/internal/client/client_loadreport_test.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+	if len(receivedLoad) <= 0 {
+		t.Fatalf("unexpected load received, want load for cluster, eds, dropped for test")
+	}
+	receivedLoad[0].LoadReportInterval = nil
+	if want := (&endpointpb.ClusterStats{
+		ClusterName:          "cluster",
+		ClusterServiceName:   "eds",
+		TotalDroppedRequests: 1,
+		DroppedRequests:      []*endpointpb.ClusterStats_DroppedRequests{{Category: "test", DroppedCount: 1}},
+	}); !proto.Equal(want, receivedLoad[0]) {
```
Comment (easwars) [inline xds/internal/client/client_loadreport_test.go:137]: Would it be possible to use `protocmp.Diff` or whatever is the equivalent in the protobuf package and pass it to `cmp.Diff` so that we can print a useful error message.

File: xds/internal/client/client_test.go [hunk 12]
```diff
@@ -115,7 +115,7 @@ func (c *testAPIClient) RemoveWatch(resourceType ResourceType, resourceName stri
 	c.removeWatches[resourceType].Send(resourceName)
 }
 
-func (c *testAPIClient) ReportLoad(ctx context.Context, cc *grpc.ClientConn, opts LoadReportingOptions) {
+func (c *testAPIClient) reportLoad(_ context.Context, _ *grpc.ClientConn, _ loadReportingOptions) {
```
Comment (easwars) [inline xds/internal/client/client_test.go:118]: The underscores could be removed I think.

File: xds/internal/balancer/edsbalancer/xds_client_wrapper.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (menghanl) [reply xds/internal/balancer/edsbalancer/xds_client_wrapper.go]: That sounds better. Done.

File: xds/internal/balancer/lrs/balancer.go (hunk 7 above)
Comment (menghanl) [reply xds/internal/balancer/lrs/balancer.go]: Added `lrs:`.

The caller (which is the implementation of the balancer interface) could wrap. But that will be only to add the prefix, and seems to be not worth it. I added it here.

File: xds/internal/balancer/lrs/balancer.go (hunk 6 above)
Comment (menghanl) [reply xds/internal/balancer/lrs/balancer.go]: This whole if statement is deleted.

File: xds/internal/client/client_loadreport.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (menghanl) [reply xds/internal/client/client_loadreport.go:126]: I didn't want to add panic...

What I meant was, this should never happen (for real????). But we still `Info` just in case..

File: xds/internal/client/client_loadreport_test.go (hunk 9 above)
Comment (menghanl) [reply xds/internal/client/client_loadreport_test.go]: Done

File: xds/internal/client/client_loadreport_test.go (hunk 10 above)
Comment (menghanl) [reply xds/internal/client/client_loadreport_test.go]: Done

File: xds/internal/client/client_loadreport_test.go (hunk 11 above)
Comment (menghanl) [reply xds/internal/client/client_loadreport_test.go]: Done.

File: xds/internal/client/client_test.go (hunk 12 above)
Comment (menghanl) [reply xds/internal/client/client_test.go:118]: Done

File: xds/internal/balancer/edsbalancer/eds.go [hunk 13]
```diff
@@ -177,7 +177,9 @@ func (x *edsBalancer) handleGRPCUpdate(update interface{}) {
 			return
 		}
 
-		x.client.handleUpdate(cfg, u.ResolverState.Attributes)
+		if err := x.client.handleUpdate(cfg, u.ResolverState.Attributes); err != nil {
+			x.logger.Infof("failed to update xds clients: %v", err)
```
Comment (easwars) [inline xds/internal/balancer/edsbalancer/eds.go:181]: `Errorf` or `Warningf` instead?

File: xds/internal/balancer/edsbalancer/xds_client_wrapper.go [hunk 14]
```diff
@@ -48,44 +48,75 @@ var (
...
+	if lsw.store != nil {
+		lsw.perCluster = lsw.store.PerCluster(lsw.service, "")
+	}
+
+}
+
 func (lsw *loadStoreWrapper) CallStarted(locality string) {
 	lsw.mu.RLock()
 	defer lsw.mu.RUnlock()
+	if lsw.perCluster == nil {
```
Comment (easwars) [inline xds/internal/balancer/edsbalancer/xds_client_wrapper.go:90]: The `perClusterStore` type in package `load` already checks for `nil`. Why is this required?

File: xds/internal/balancer/edsbalancer/xds_client_wrapper.go [hunk 15]
```diff
@@ -184,24 +217,28 @@ func (c *xdsClientWrapper) updateXDSClient(config *EDSConfig, attr *attributes.A
...
-		return false
+		return false, nil
 	}
 
 	var dopts []grpc.DialOption
 	if dialer := c.bbo.Dialer; dialer != nil {
 		dopts = []grpc.DialOption{grpc.WithContextDialer(dialer)}
 	}
 
+	// TODO: there's no long a need to read bootstrap file and create a new xds
```
Comment (easwars) [inline xds/internal/balancer/edsbalancer/xds_client_wrapper.go:228]: s/long/longer/

File: xds/internal/balancer/lrs/balancer.go [hunk 16]
```diff
@@ -144,52 +146,81 @@ func (ccw *ccWrapper) UpdateState(s balancer.State) {
...
+	lsw.perCluster = nil
+	if lsw.store != nil {
+		lsw.perCluster = lsw.store.PerCluster(lsw.cluster, lsw.edsService)
+	}
+}
+
 func (lsw *loadStoreWrapper) CallStarted(locality string) {
 	lsw.mu.RLock()
 	defer lsw.mu.RUnlock()
+	if lsw.perCluster == nil {
```
Comment (easwars) [inline xds/internal/balancer/lrs/balancer.go:194]: Same comment about `nil` checks.

File: xds/internal/balancer/edsbalancer/eds.go (hunk 13 above)
Comment (menghanl) [reply xds/internal/balancer/edsbalancer/eds.go]: Changed to warning

File: xds/internal/balancer/edsbalancer/xds_client_wrapper.go (hunk 14 above)
Comment (menghanl) [reply xds/internal/balancer/edsbalancer/xds_client_wrapper.go]: Right, I forgot.
Deleted.

File: xds/internal/balancer/edsbalancer/xds_client_wrapper.go (hunk 15 above)
Comment (menghanl) [reply xds/internal/balancer/edsbalancer/xds_client_wrapper.go]: Done

File: xds/internal/balancer/lrs/balancer.go (hunk 16 above)
Comment (menghanl) [reply xds/internal/balancer/lrs/balancer.go]: Removed

**Review by easwars — APPROVED** [review]
